subsys: fb: cfb: support drawing to any coordinates

soburi opened the pull request:
This PR intends to use cfb like as a graphics oriented devices. It will be fundamental for implements simple graphics rendering.

# API improvement:

- Allow printing and inverting at any coordinates.

  Currently, the CFB API can only draw at tile bounds coordinates (where the y coordinate is a multiple of 8).
  Implements dot-by-dot font rendering to make cfb_print() able to draw text even at a coordinate that is not on tile boundaries.
  For API consistency, change cfb_invert_area() able to invert even at a coordinate not on tile boundaries.

- Add `cfb_draw_text()` API

  Added `cfb_draw_text()` that does not wrap when specifying the position in dots, as it may be inconvenient if there is wrapping

# Shell Improvements:

- Add command `draw text`

  Add command to run the new API `cfb_draw_text()`.
  This command is similar to `print`.
  The position specify with (x,y) not column-row and doesn't clear buffer is differences.

- Add support for running cfb_invert_area().
  
  Extend the invert command to run cfb_invert_area when given four arguments.
  Make a change to invert the screen immediately in case of run invert without arguments.

soburi committed:
fb: cfb: support drawing to any coordinates

Implements dot-by-dot font rendering to make cfb_print() able to
draw text even at a coordinate that is not on tile boundaries.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
fb: cfb: support inverting with coordinates that do not align with the tile

Improve `cfb_invert_area()` able to invert even at a coordinate
not on tile boundaries.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
fb: cfb_shell: support area inverting

Extend the invert command to run cfb_invert_area when given four arguments.
Make a change to invert the screen immediately in case of run invert
without arguments.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
fb: cfb_shell: Add `draw text` command

Add the `draw text` command to execute the `cfb_draw_text()` API.

The command is similar to the `print` command. The difference is
`draw text` can render to coordinate that is not on tile boundaries.
The command is not run `cfb_framebuffer_clear()`,
So it allows rendering appendiceal with keeping already drawn image.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
fb: cfb: Restore inverted framebuffer to make able to reuse

cfb_framebuffer_finalize() invert the framebuffer, and it still
leave as inverted after executing the function.

It restores the framebuffer at the end of the cfb_framebuffer_finalize()
for the continued drawing.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
fb: cfb: Add cfb_draw_text() API

Add cfb_draw_text() API for rendering text.
It is similar to cfb_print(), the difference is cfb_draw_text() accepts
coordinate that is not on tile boundaries and don't wrap the line.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

MaureenHelm commented:
@jfischer-no will you please have a look?

jfischer-no commented on subsys/fb/cfb.c:
```diff
@@ -105,24 +105,39 @@ static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
+
+		for (size_t g_y = 0; g_y < fptr->height; g_y++) {
+			int16_t fb_y = y + g_y;
+			size_t fb_index = (fb_y / 8U) * fb->x_res + fb_x;
```
Would be nice to have short and concise comments on few points, otherwise no one can understand what happens at first glance.

soburi replied:
Add explaining the purpose of each in the process branching.

soburi commented on subsys/fb/cfb.c:
```diff
@@ -87,8 +87,8 @@ static inline uint8_t get_glyph_byte(uint8_t *glyph_ptr, const struct cfb_font *
  */
-static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
-			     char c, uint16_t x, uint16_t y)
+static uint8_t draw_char_vtmono(const struct char_framebuffer *fb, char c, uint16_t x, uint16_t y,
```
Thank you for reviewing.

Align the line start.

soburi commented on subsys/fb/cfb.c:
```diff
@@ -105,24 +105,39 @@ static uint8_t draw_char_vtmono(const struct char_framebuffer *fb,
+				}
+				g_y += 7;
+			} else {
+				if (byte & BIT(g_y % 8)) {
```
It make a sense.

I rewrite this code to run less times loop.

5frank commented on include/zephyr/display/cfb.h:
```diff
@@ -89,6 +89,20 @@ struct cfb_font {
+ *
+ * @return 0 on success, negative value otherwise
+ */
+int cfb_draw_text(const struct device *dev, char *str, int16_t x, int16_t y);
```
Would be nice with `const char *str`!?  `cfb_print` should probably also have const parameter IMO.

soburi committed:
fb: cfb: add const modifier for argument of cfb_print

cfb_print does not modify either address and contents of
the pointer that is pointing drawing text.
Thus, it can add a const modifier.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

teburd commented:
@jfischer-no @5frank please look again

soburi commented:
@5frank 
ping

MaureenHelm commented:
@fabiobaltieri please take a look
